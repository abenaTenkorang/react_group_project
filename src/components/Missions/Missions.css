.missionTable {
  border-collapse: collapse;
}

.missionTable thead {
  background-color: #ddd;
}

.missionTable tr:nth-child(even) {
  background-color: #eee;
}

.missionTable th,
.missionTable td {
  border: 1px solid #ccc;
  padding: 0.5rem;
}

.status-badge {
  background: #999;
  color: white;
  font-weight: bold;
  border-radius: 0.25rem;
  display: block;
  text-align: center;
  padding: 0.15rem;
  font-size: 0.65rem;
  text-transform: uppercase;
  min-width: 8rem;
}

.status-badge.joined {
  background-color: lightseagreen;
}

.action-button {
  border: 1px solid #777;
  border-radius: 0.5rem;
  display: block;
  text-align: center;
  padding: 0.25rem;
  font-size: 0.85rem;
  margin-top: 1rem;
  cursor: pointer;
  min-width: 8rem;
}

.action-button.join-button {
  color: green;
  border-color: green;
}

.action-button.leave-button {
  color: red;
  border-color: red;
}
