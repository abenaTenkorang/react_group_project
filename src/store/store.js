import { configureStore } from '@reduxjs/toolkit';
import rocketReducer from './rocket/rocketSlice';
import missionsReducer from './missions/missions';

export default configureStore({
  reducer: {
    rockets: rocketReducer,
    missions: missionsReducer,
  },
});
