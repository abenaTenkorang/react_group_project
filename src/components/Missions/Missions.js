import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  fetchMissionsAsync,
  joinMission,
  leaveMission,
} from "../../store/missions/missions";
import "./Missions.css";

export default function Missions() {
  const missions = useSelector((state) => state.missions.missions || []);
  const loading = useSelector((state) => state.missions.isLoading);
  const error = useSelector((state) => state.missions.error);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!missions.length) dispatch(fetchMissionsAsync());
  }, [dispatch, missions.length]);

  return (
    <section className="container">
      <h2>Missions</h2>
      {error && (
        <p>
          Failed to fetch missions:
          {error}
        </p>
      )}
      {!loading && !missions.length && <p>No missions found</p>}
      {loading && <p>Loading missions...</p>}
      {Array.isArray(missions) && missions.length > 0 && (
        <table className="missionTable">
          <thead>
            <tr>
              <th>Name</th>
              <th>Description</th>
              <th>Status</th>
            </tr>
          </thead>
          <tbody>
            {missions.map((mission) => (
              <tr key={mission.id}>
                <td>{mission.name}</td>
                <td className="font-sm">{mission.description}</td>
                <td className="mission-status-cell">
                  {!mission.joined && (
                    <>
                      <span className="status-badge">Not a member</span>
                      <button
                        type="button"
                        className="action-button join-button"
                        onClick={() => dispatch(joinMission(mission.id))}
                      >
                        Join Mission
                      </button>
                    </>
                  )}
                  {mission.joined && (
                    <>
                      <span className="status-badge joined">Active Member</span>
                      <button
                        type="button"
                        className="action-button leave-button"
                        onClick={() => dispatch(leaveMission(mission.id))}
                      >
                        Leave Mission
                      </button>
                    </>
                  )}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      )}
    </section>
  );
}
