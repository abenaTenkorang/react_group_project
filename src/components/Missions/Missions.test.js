import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../store/store';
import Missions from './Missions';

describe('Missions', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Missions />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
