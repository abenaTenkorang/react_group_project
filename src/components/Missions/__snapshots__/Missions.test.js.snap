// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions renders correctly 1`] = `
<section
  className="container"
>
  <h2>
    Missions
  </h2>
  <p>
    Loading missions...
  </p>
</section>
`;
